packages/core/src/BlockNoteEditor.ts

```diff
@@ -159,19 +159,30 @@ export class BlockNoteEditor {
    * @param reverse Whether the blocks should be traversed in reverse order.
    */
   public forEachBlock(
-    callback: (block: Block) => void,
+    callback: (block: Block) => boolean,
     reverse: boolean = false
   ): void {
+    let keepTraversing = true;
     function helper(blocks: Block[]) {
       if (reverse) {
         for (const block of blocks.reverse()) {
-          helper(block.children);
-          callback(block);
+          if (keepTraversing) {
+            helper(block.children);
+          }
+
+          if (keepTraversing) {
```

Comment:
I don't think so, since we'll only return at that recursive branch & depth. That is what ProseMirror does with `descendants` but that only stops recursing deeper into the same branch, whereas this approach stops the traversal entirely. Could do it either way, imo this way has a broader use case though.

Reply:
After going back and forth with chatgpt I got this:

```
/**
   * Traverses all blocks in the editor depth-first, and executes a callback for each.
   * @param callback The callback to execute for each block. Returning `false` stops the traversal.
   * @param reverse Whether the blocks should be traversed in reverse order.
   */
  public forEachBlock(
    callback: (block: Block) => boolean,
    reverse: boolean = false
  ): void {
    const blocks = this.topLevelBlocks.slice();

    if (reverse) {
      blocks.reverse();
    }

    function traverseBlockArray(blockArray: Block[]): boolean {
      for (const block of blockArray) {
        if (callback(block) === false) {
          return false;
        }

        const children = reverse ? block.children.slice().reverse() : block.children;

        if (traverseBlockArray(children) === false) {
          return false;
        }
      }

      return true;
    }

    traverseBlockArray(blocks);
  }

```

wdyt? Important change is that `.reverse()` modifies an array in-place, so the `slice` is added

Reply:
Yeah this looks good